server/service/src/processors/requisition_auto_finalise/requisition_auto_finalise.rs

```diff
@@ -0,0 +1,178 @@
+            .ok_or(ProcessorError::RecordNotFound(
+                "Requisition".to_string(),
+                changelog.record_id.clone(),
+            ))?;
+
+        if requisition.r#type != RequisitionType::Response {
+            return Ok(None);
+        }
+
+        if requisition.status == RequisitionStatus::Finalised {
+            return Ok(None);
+        }
+
+        let invoices = InvoiceRepository::new(connection).query_by_filter(
+            InvoiceFilter::new().requisition_id(EqualFilter::equal_to(&requisition.id)),
+        )?;
+
+        let requisition_lines = RequisitionLineRepository::new(connection).query_by_filter(
+            RequisitionLineFilter::new().requisition_id(EqualFilter::equal_to(&requisition.id)),
+        )?;
+        if requisition_lines.len() == 0 {
+            return Ok(None);
+        }
+
+        let invoice_lines = InvoiceLineRepository::new(connection).query_by_filter(
+            InvoiceLineFilter::new()
+                .invoice_id(EqualFilter::equal_any(
+                    invoices.into_iter().map(|i| i.invoice_row.id).collect(),
+                ))
+                .r#type(EqualFilter {
+                    equal_to: Some(InvoiceLineType::StockOut),
+                    ..Default::default()
+                }),
+        )?;
+        if invoice_lines.len() == 0 {
+            return Ok(None);
+        }
+
+        let mut invoice_line_item_units: HashMap<String, f64> = HashMap::new();
+        invoice_lines
+            .iter()
+            .filter_map(|il| {
+                if matches!(
+                    il.invoice_row.status,
+                    InvoiceStatus::Allocated | InvoiceStatus::New | InvoiceStatus::Picked
+                ) {
+                    return None;
+                }
+                Some(il)
+            })
+            .for_each(|il| {
+                let units = il.invoice_line_row.number_of_packs * il.invoice_line_row.pack_size;
+                invoice_line_item_units
+                    .entry(il.item_row.id.clone())
+                    .and_modify(|v| *v += units)
+                    .or_insert(units);
+            });
+        let should_finalise = requisition_lines.iter().all(|rl| {
+            if rl.requisition_line_row.supply_quantity <= 0.0 {
+                return true;
+            }
+            invoice_line_item_units
+                .get(&rl.item_row.id)
+                .is_some_and(|v| *v >= rl.requisition_line_row.supply_quantity)
+        });
+
+        if !should_finalise {
+            return Ok(None);
+        }
+
+        requisition.status = RequisitionStatus::Finalised;
+        requisition.finalised_datetime = Some(chrono::Utc::now().naive_utc());
+
+        requisition_row_repo.upsert_one(&requisition)?;
+
+        system_activity_log_entry(
+            connection,
+            ActivityLogType::RequisitionStatusFinalised,
+            &requisition.store_id,
+            &requisition.id,
+        )?;
```

Comment:
Probably better to use `update_response_requisition`? The validation there might be important and if more statuses are added it'll probably handle setting multiple datetime fields (and this file neglected).

That said, I'd like the log to say something along the lines of "Requisition auto finalised" in its detail. I haven't done that here, and it's even harder with `update_response_requisition`... well at least not without making the function parameters ugly.